forc-pkg/src/source/reg/mod.rs

```diff
@@ -0,0 +1,531 @@
+}
+
+impl From<Pinned> for source::Pinned {
+    fn from(p: Pinned) -> Self {
+        Self::Registry(p)
+    }
+}
+
+async fn fetch(fetch_id: u64, pinned: &Pinned, ipfs_node: &IPFSNode) -> anyhow::Result<PathBuf> {
+    let path = with_tmp_fetch_index(
+        fetch_id,
+        &pinned.source.name,
+        &pinned.source,
+        |index_file| async move {
+            let path = registry_package_dir(
+                &pinned.source.namespace,
+                &pinned.source.name,
+                &pinned.source.version,
+            );
+            if path.exists() {
+                let _ = fs::remove_dir_all(&path);
+            }
+            fs::create_dir_all(&path)?;
+
+            let package_entry = index_file.get(&pinned.source.version).ok_or_else(|| {
+                anyhow!(
+                    "Version {} not found for {}",
+                    pinned.source.version,
+                    pinned.source.name
+                )
+            })?;
+
+            let cid = Cid::from_str(&package_entry.source_cid).with_context(|| {
+                format!(
+                    "Invalid CID {}v{}: `{}`",
+                    package_entry.name, package_entry.version, package_entry.source_cid
+                )
+            })?;
+            match ipfs_node {
+                IPFSNode::Local => {
+                    println_action_green("Fetching", "with local IPFS node");
+                    cid.fetch_with_client(&ipfs_client(), &path).await?;
+                }
+                IPFSNode::WithUrl(gateway_url) => {
+                    println_action_green(
+                        "Fetching",
+                        &format!("from {}. Note: This can take several minutes.", gateway_url),
+                    );
+                    cid.fetch_with_gateway_url(gateway_url, &path).await?;
+                }
+            }
+
+            Ok(path)
+        },
+    )
+    .await?;
+    Ok(path)
+}
+
+async fn with_tmp_fetch_index<F, O, Fut>(
+    fetch_id: u64,
+    pkg_name: &str,
+    source: &Source,
+    f: F,
+) -> anyhow::Result<O>
+where
+    F: FnOnce(IndexFile) -> Fut,
+    Fut: std::future::Future<Output = anyhow::Result<O>>,
+{
+    let tmp_dir = tmp_registry_package_dir(fetch_id, pkg_name, &source.version, &source.namespace);
+    if tmp_dir.exists() {
+        let _ = std::fs::remove_dir_all(&tmp_dir);
+    }
+
+    let github_resolver = GithubRegistryResolver::with_default_github(source.namespace.clone());
+
+    let path = format!(
+        "{}",
+        location_from_root(github_resolver.chunk_size, &source.namespace, pkg_name).display()
+    );
+    let index_repo_owner = github_resolver.repo_org();
+    let index_repo_name = github_resolver.repo_name();
+    let reference = format!("refs/heads/{}", github_resolver.branch_name());
+    let github_endpoint = format!(
+        "https://raw.githubusercontent.com/{index_repo_owner}/{index_repo_name}/{reference}/{path}"
+    );
+    let client = reqwest::Client::new();
```

Comment:
should we add a timeout here to prevent it hanging?

Reply:
Sure, done in [a834871](https://github.com/FuelLabs/sway/pull/7038/commits/a834871f5a097a293ea5bc053229d0ab8eccf566)